### PR TITLE
[SPARK-47754][SQL] Postgres: Support reading multidimensional arrays

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Because the ResultSetMetadata cannot distinguish a single-dimensional array from multidimensional arrays. Thus, we always read multidimensional arrays as single-dimensional ones, For example, `text[][]` is mapping to `ArrayType(StringType)` and `int[][][]` is `ArrayType(IntegerType)`, this result in errors when converting a ResultSet with multidimensional arrays to InternalRows.

This PR supports reading multidimensional arrays from PostgreSQL data sources. To achieve this, the simplest way is to add a new developer API to retrieve it from the information schema of Postgres.

https://www.postgresql.org/docs/16/catalog-pg-attribute.html#CATALOG-PG-ATTRIBUTE

It is possible to use functions like `array_dims` to retrieve the dimension of an array column, but it is not easy to inject without causing breaking changes or to determine the dimension based on the actual data.
### Why are the changes needed?

We have supported writing multidimensional arrays to Postgres, so we shall improve postgres reading abilities too.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @dongjoon-hyun thanks

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 1]
```diff
@@ -798,6 +798,9 @@ abstract class JdbcDialect extends Serializable with Logging {
   protected final def getTimestampType(md: Metadata): DataType = {
     JdbcUtils.getTimestampType(md.getBoolean("isTimestampNTZ"))
   }
+
+  @Since("4.0.0")
+  def getArrayDimension(conn: Connection, tableName: String, columnName: String): Option[Int] = None
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:803]: Could you add a function description to describe the meaning of return value? For example, in the following format
- Some(n): What is the definition of n
- None: What is the meaning

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with one minor comment about function description)

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:816]: Thank you for the suggestion @dongjoon-hyun. I have updated the desc.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you!

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun

**Pull request closed without merging**